cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go

```diff
@@ -0,0 +1,361 @@
+}
+
+// nodeGroupSize gets the current size of the nodegroup as reported by packet tags.
+func (mgr *packetManagerRest) nodeGroupSize(nodegroup string) (int, error) {
+	devices := mgr.listPacketDevices()
+	// Get the count of devices tagged as nodegroup members
+	count := 0
+	for _, d := range devices.Devices {
+		if Contains(d.Tags, "k8s-cluster-"+mgr.clusterName) && Contains(d.Tags, "k8s-nodepool-"+nodegroup) {
+			count++
+		}
+	}
+	fmt.Println(len(devices.Devices), count)
+	return count, nil
+}
+
+func randString8() string {
+	n := 8
+	rand.Seed(time.Now().UnixNano())
+	letterRunes := []rune("acdefghijklmnopqrstuvwxyz")
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letterRunes[rand.Intn(len(letterRunes))]
+	}
+	return string(b)
+}
+
+// createNodes provisions new nodes on packet and bootstraps them in the cluster.
+func (mgr *packetManagerRest) createNodes(nodegroup string, nodes int) error {
+	klog.Infof("Updating node count to %d for nodegroup %s", nodes, nodegroup)
+	cloudinit, err := base64.StdEncoding.DecodeString(mgr.cloudinit)
+	if err != nil {
+		log.Fatal(err)
+		return fmt.Errorf("Could not decode cloudinit script: %v", err)
+	}
+
+	udvars := CloudInitTemplateData{
+		BootstrapTokenID:     os.Getenv("BOOTSTRAP_TOKEN_ID"),
+		BootstrapTokenSecret: os.Getenv("BOOTSTRAP_TOKEN_SECRET"),
+		APIServerEndpoint:    mgr.apiServerEndpoint,
+	}
+	ud := renderTemplate(string(cloudinit), udvars)
+	hn := "k8s-" + mgr.clusterName + "-" + nodegroup + "-" + randString8()
+
+	reservation := ""
+	if mgr.reservation == "require" || mgr.reservation == "prefer" {
+		reservation = "next-available"
+	}
+
+	cr := DeviceCreateRequest{
+		Hostname:              hn,
+		Facility:              []string{mgr.facility},
+		Plan:                  mgr.plan,
+		OS:                    mgr.os,
+		ProjectID:             mgr.projectID,
+		BillingCycle:          mgr.billing,
+		UserData:              renderTemplate(ud, udvars),
+		Tags:                  []string{"k8s-cluster-" + mgr.clusterName, "k8s-nodepool-" + nodegroup},
+		HardwareReservationID: reservation,
+	}
+
+	resp, err := createDevice(&cr)
+	if err != nil || resp.StatusCode > 299 {
+		// If reservation is preferred but not available, retry provisioning as on-demand
+		if reservation != "" && mgr.reservation == "prefer" {
+			klog.Infof("Reservation preferred but not available. Provisioning on-demand node.")
+			cr.HardwareReservationID = ""
+			resp, err = createDevice(&cr)
+			if err != nil {
+				klog.Errorf("Failed to create device using Packet API: %v", err)
+				panic(err)
+			} else if resp.StatusCode > 299 {
+				klog.Errorf("Failed to create device using Packet API: %v", resp)
+				panic(resp)
+			}
+		} else if err != nil {
+			klog.Errorf("Failed to create device using Packet API: %v", err)
+			panic(err)
+		} else if resp.StatusCode > 299 {
+			klog.Errorf("Failed to create device using Packet API: %v", resp)
+			panic(resp)
+		}
+	}
+	defer resp.Body.Close()
+
+	ioutil.ReadAll(resp.Body)
```

Comment:
Maybe a comment or assign to `_` so it is clear it is ignored?

Reply:
Right, fixed. Also added some verbose logging of the response and some error handling.